crates/rpc/src/server/accept.rs

```diff
@@ -0,0 +1,188 @@
+                    Ok(accept) => match Version::parse(accept.version) {
+                        // Verify the version matches configured requirements.
+                        Ok(version) => {
+                            if self.version.matches(&version) {
+                                self.inner.call(request).boxed()
+                            } else {
+                                debug!(target: COMPONENT, "Version does not match ({}/{})", version, self.version);
+                                bad_request("Client / server version mismatch").boxed()
+                            }
+                        },
+                        Err(e) => {
+                            debug!(target: COMPONENT, "Failed to parse version: {}", e);
+                            bad_request("Invalid version specified in accept header value").boxed()
+                        },
+                    },
+                    Err(e) => {
+                        debug!(target: COMPONENT, "Failed to parse accept header value: {}", e);
+                        bad_request("Invalid accept header value").boxed()
+                    },
+                },
+                Err(e) => {
+                    debug!(target: COMPONENT, "Failed to stringify accept header value: {}", e);
+                    bad_request("Invalid accept header value").boxed()
+                },
+            }
+        } else {
+            debug!(target: COMPONENT, "Request missing ACCEPT header");
+            bad_request("Missing required ACCEPT header").boxed()
+        }
+    }
+}
+
+/// Returns a future that resolves to a bad request response.
+fn bad_request<B: Default + Send + 'static, E: Send + 'static>(
+    msg: &'static str,
+) -> impl Future<Output = Result<http::Response<B>, E>> {
+    let response = http::Response::builder()
+        .status(http::StatusCode::BAD_REQUEST)
+        .header("content-type", "application/grpc")
+        .header("grpc-status", "3") // INVALID_ARGUMENT
+        .header("grpc-message", msg)
+        .body(B::default())
+        .expect("headers are valid");
+
+    futures::future::ready(Ok(response))
+}
+
+/// The result of parsing the following canonical representation
+/// of an ACCEPT header value: `application/vnd.{app_name}.{version}+{response_type}`.
+#[derive(Debug, PartialEq)]
+pub struct AcceptHeaderValue<'a> {
+    app_name: &'a str,
+    version: &'a str,
+    response_type: &'a str,
+}
+
+impl AcceptHeaderValue<'_> {
+    /// Parses the given input string into an [`AcceptHeaderValue`].
+    fn parse(input: &str) -> IResult<&str, AcceptHeaderValue> {
+        let (input, _) = tag("application/vnd.")(input)?;
+        let (input, app_name) = take_until(".")(input)?;
+        let (input, _) = tag(".")(input)?;
+        let (input, version) = take_until("+")(input)?;
+        let (response_type, _) = tag("+")(input)?;
+
+        Ok(("", AcceptHeaderValue { app_name, version, response_type }))
+    }
```

Comment:
Why do we need to use `IResult` here as the return type and pass an empty string as the first parameter?

Or more generally, why not move this logic into the `try_from()` method below and get rid of `parse()` method?

Reply:
Its the idiomatic way of working with `nom` - all `nom` fns do the same.  The idea is that fns responsible for parsing return the remainder of the &str that was not consumed for the purposes of debugging / unit testing.

I have updated it as per your suggestion and removed the parse fn altogether. We have just lost access to the remainder of the parsed string for the sake of debugging. Shouldn't be an issue though.